Assets/HoloToolkit/SpatialMapping/Scripts/SpatialMappingSource.cs

```diff
@@ -70,77 +103,189 @@ protected GameObject AddSurfaceObject(Mesh mesh, string objectName, Transform pa
+                    }
+
+                    CleanUpSurface(
+                        existing,
+                        destroyGameObjectIfReplaced,
+                        destroyMeshesIfReplaced,
+                        objectToPreserve: toUpdateOrAdd.Object,
+                        meshToPreserveA: toUpdateOrAdd.Filter.sharedMesh,
+                        meshToPreserveB: toUpdateOrAdd.Collider.sharedMesh
+                        );
+
+                    replaced = existing;
+                    break;
                 }
             }
 
-            // Not in the list, add it
-            SurfaceObject surfaceObject = new SurfaceObject();
-            surfaceObject.ID = meshID;
-            surfaceObject.UpdateID = 0;
-
-            surfaceObject.Object = surfaceGameObject;
-            surfaceObject.Filter = surfaceObject.Object.GetComponent<MeshFilter>();
-            surfaceObject.Renderer = surfaceObject.Object.GetComponent<MeshRenderer>();
+            if (replaced == null)
+            {
+                AddSurfaceObject(toUpdateOrAdd);
+            }
 
-            SurfaceObjects.Add(surfaceObject);
+            return replaced;
         }
 
         /// <summary>
-        /// Removes and optionally destroys the specified surfaceObject that we've created
+        /// Remove the first surface with the specified ID if one exists in <see cref="SurfaceObjects"/>.
         /// </summary>
-        /// <param name="surface">The surface game object</param>
-        /// <param name="removeAndDestroy">If true, the surface will be removed and destroyed. If false, it will only be removed.</param>
-        protected void RemoveSurfaceObject(GameObject surface, bool removeAndDestroy = true)
+        /// <param name="surfaceID">The ID of the surface to remove.</param>
+        /// <param name="destroyGameObject">True to destroy the <see cref="SurfaceObject.Object"/> associated with the surface, false otherwise.</param>
+        /// <param name="destroyMeshes">True to destroy the meshes associated with the surface, false otherwise.</param>
+        /// <returns>The surface object if one was found and removed or null if one was not found.</returns>
+        protected SurfaceObject? RemoveSurfaceIfFound(int surfaceID, bool destroyGameObject = true, bool destroyMeshes = true)
         {
-            // Remove it from our list
-            for (int index = 0; index < SurfaceObjects.Count; index++)
+            SurfaceObject? removed = null;
+
+            for (int iSurface = 0; iSurface < surfaceObjectsWriteable.Count; iSurface++)
             {
-                if (SurfaceObjects[index].Object == surface)
+                var surface = surfaceObjectsWriteable[iSurface];
+
+                if (surface.ID == surfaceID)
                 {
-                    if (removeAndDestroy)
+                    surfaceObjectsWriteable.RemoveAt(iSurface);
+
+                    var handlers = SurfaceRemoved;
```

Comment:
This one I'd like to keep as "var".  Same rationale as the other "handlers" one.